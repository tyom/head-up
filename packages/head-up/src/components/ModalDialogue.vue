<template>
  <transition name="reveal">
    <div v-if="show" class="ModalDialogue">
      <div class="backdrop" @click="$emit('close')"/>
      <div class="container">
        <header v-if="heading">
          <h2 class="title">{{ heading }}</h2>
        </header>
        <slot/>
      </div>
    </div>
  </transition>
</template>

<script>
export default {
  props: {
    show: {
      type: Boolean,
      default: false,
    },
    heading: {
      type: String,
      default: '',
    },
  },
};
</script>

<style scoped>
.ModalDialogue {
  position: fixed;
  z-index: 2;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.backdrop {
  background-color: #000b;
  height: 100%;
  width: 100%;
}

.container {
  position: absolute;
  border-radius: 0.3em;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  max-width: 60vw;
  max-height: 60vh;
  margin: auto;
  z-index: 1;
  background-color: #fff;
  color: #444;
  padding: 1em;
}

header {
  padding-bottom: 1em;
  margin-bottom: 1em;
  border-bottom: 0.1em solid #ccc;
}

.reveal-enter-active,
.reveal-leave-active {
  transition: 1s;
}
.reveal-enter-active {
  & .backdrop {
    transition: 0.5s;
  }

  & .container {
    transition: 0.2s 0.1s;
  }
}
.reveal-leave-active {
  & .backdrop {
    transition: 0.3s 0.1s;
  }

  & .container {
    transition: 0.1s;
  }
}

.reveal-enter,
.reveal-leave-to {
  & .backdrop {
    opacity: 0;
  }

  & .container {
    opacity: 0;
    transform: translateY(-50%);
  }
}
</style>
