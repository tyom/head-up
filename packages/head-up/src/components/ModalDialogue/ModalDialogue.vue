<template>
  <transition name="reveal">
    <div
      class="ModalDialogue"
      v-shortkey="['esc']"
      @shortkey="handleClose"
    >
      <div class="backdrop" @click="handleClose"/>
      <div class="container">
        <header v-if="heading">
          <h2 class="title">{{ heading }}</h2>
        </header>
        <slot/>
      </div>
    </div>
  </transition>
</template>

<script>
export default {
  props: {
    heading: {
      type: String,
      default: '',
    },
  },
  methods: {
    handleClose() {
      this.$emit('close');
    },
  },
};
</script>

<style scoped>
.ModalDialogue {
  position: fixed;
  z-index: 2;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.backdrop {
  background-color: #000b;
  height: 100%;
  width: 100%;
}

.container {
  position: absolute;
  border-radius: 0.3em;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  max-width: 60vw;
  max-height: 60vh;
  margin: auto;
  z-index: 1;
  background-color: #222;
  border: 2px solid #fff4;
  color: #ccc;
  padding: 1em;
}

header {
  padding: 1em;
  margin: -1em -1em 1em;
  background-color: #0002;
  border-bottom: 0.1em solid #fff2;
}

.reveal-enter-active,
.reveal-leave-active {
  transition: 1s;
}
.reveal-enter-active {
  & .backdrop {
    transition: 0.5s;
  }

  & .container {
    transition: 0.2s 0.1s;
  }
}
.reveal-leave-active {
  & .backdrop {
    transition: 0.3s 0.1s;
  }

  & .container {
    transition: 0.1s;
  }
}

.reveal-enter,
.reveal-leave-to {
  & .backdrop {
    opacity: 0;
  }

  & .container {
    opacity: 0;
    transform: translateY(-50%);
  }
}
</style>
