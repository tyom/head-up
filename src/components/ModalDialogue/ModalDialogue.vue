<template>
  <transition name="reveal">
    <div
      v-if="content"
      class="ModalDialogue"
      v-shortkey="['esc']"
      @shortkey="handleClose"
    >
      <div class="backdrop" @click="handleClose" />
      <div class="wrapper">
        <header v-if="content.heading">
          <h2 class="title">{{ content.heading }}</h2>
        </header>
        <div class="body">
          <component :is="modalComponent" />
        </div>
      </div>
    </div>
  </transition>
</template>

<script>
import SettingsScreen from '../SettingsScreen';
import HelpScreen from '../HelpScreen';

export default {
  components: {
    SettingsScreen,
    HelpScreen,
  },
  props: {
    content: {
      type: Object,
      default: null,
    },
  },
  computed: {
    modalComponent() {
      if (!this.content) {
        return;
      }
      switch (this.content.name) {
        case 'help':
          return 'HelpScreen';
        case 'settings':
          return 'SettingsScreen';
        default:
          return '';
      }
    },
  },
  methods: {
    handleClose() {
      this.$emit('close');
    },
  },
};
</script>

<style scoped>
.ModalDialogue {
  @apply fixed inset-0 z-10 flex justify-center items-center text-lg;
}

.backdrop {
  @apply h-full w-full;
  background-color: #000b;
}

.wrapper {
  @apply absolute rounded-md border-solid border-2 shadow-lg;
  @apply bg-gray-800 border-gray-600 text-gray-400 transition;
}

header {
  @apply p-4 text-xl border-b-2 border-gray-600;
}

.body {
  @apply p-4 overflow-auto;
  width: 60vw;
  min-height: 30vh;
}

.reveal-enter-active,
.reveal-leave-active {
  @apply transition duration-700;
}
.reveal-enter-active {
  & .backdrop {
    @apply duration-500;
  }

  & .wrapper {
    transition: 0.2s 0.1s;
  }
}
.reveal-leave-active {
  & .backdrop {
    transition: 0.3s 0.1s;
  }

  & .wrapper {
    @apply duration-100;
  }
}

.reveal-enter,
.reveal-leave-to {
  & .backdrop {
    @apply opacity-0;
  }

  & .wrapper {
    @apply opacity-0 transform -translate-y-20;
  }
}
</style>
