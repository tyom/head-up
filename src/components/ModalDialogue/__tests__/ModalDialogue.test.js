import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import ShortKey from 'vue-shortkey';
import ModalDialogue from '../ModalDialogue';

const localVue = createLocalVue();

localVue.use(ShortKey);

test('render default', () => {
  const wrapper = shallowMount(ModalDialogue, {
    localVue,
  });
  expect(wrapper).toMatchSnapshot();
});

test('render help modal', () => {
  const wrapper = shallowMount(ModalDialogue, {
    localVue,
    propsData: {
      content: {
        name: 'help',
        heading: 'Help is here',
      },
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render settings modal', () => {
  const wrapper = shallowMount(ModalDialogue, {
    localVue,
    propsData: {
      content: {
        name: 'settings',
        heading: 'Settings is here',
      },
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('closes dialogue on backdrop click', () => {
  const wrapper = shallowMount(ModalDialogue, {
    localVue,
    propsData: {
      content: {
        name: 'settings',
      },
    },
  });

  wrapper.find('.backdrop').trigger('click');
  expect(wrapper.emitted('close')).toBeTruthy();
});

test('closes dialogue on escape key', () => {
  const wrapper = mount(ModalDialogue, {
    attachToDocument: true,
    localVue,
    propsData: {
      content: {
        name: 'settings',
      },
    },
  });

  wrapper.trigger('keydown.esc', {
    key: 'Escape',
  });
  expect(wrapper.emitted('close')).toBeTruthy();
});
