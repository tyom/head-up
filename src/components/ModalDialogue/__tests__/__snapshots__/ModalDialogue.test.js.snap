// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<transition-stub name="reveal">
  <!---->
</transition-stub>
`;

exports[`render help modal 1`] = `
<transition-stub name="reveal">
  <div class="ModalDialogue">
    <div class="backdrop"></div>
    <div class="wrapper">
      <header>
        <h2 class="title">Help is here</h2>
      </header>
      <div class="body">
        <helpscreen-stub></helpscreen-stub>
      </div>
    </div>
  </div>
</transition-stub>
`;

exports[`render settings modal 1`] = `
<transition-stub name="reveal">
  <div class="ModalDialogue">
    <div class="backdrop"></div>
    <div class="wrapper">
      <header>
        <h2 class="title">Settings is here</h2>
      </header>
      <div class="body">
        <settingsscreen-stub></settingsscreen-stub>
      </div>
    </div>
  </div>
</transition-stub>
`;
